packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -106,87 +163,200 @@ export class BlobManager {
+    private async onUploadReject(localId: string, error) {
+        const entry = this.pendingBlobs.get(localId);
+        assert(!!entry, "Must have pending blob entry for blob which failed to upload");
+        if (!this.runtime.connected) {
+            entry.status = PendingBlobStatus.OfflinePendingUpload;
+            // since we are not connected, we will have a chance to add the storage ID when reSubmit() is called
+            this.sendBlobAttachOp(undefined, localId);
+            entry.handleP.resolve(this.getBlobHandle(localId));
+            // we are probably not connected to storage but start another upload request in case we are
+            const delay = this.retryThrottler.getDelay();
+            entry.uploadP = new Promise<void>((res) => setTimeout(res, delay))
+                .then(async () => this.uploadBlob(localId, entry.blob));
+            return entry.uploadP;
+        } else {
+            entry.handleP.reject(error);
+            throw error;
         }
+    }
 
-        return handle;
+    /**
+     * Resubmit a BlobAttach op. Used to add storage IDs to ops that were
+     * submitted to runtime while disconnected.
+     * @param metadata - op metadata containing storage and/or local IDs
+     */
+    public reSubmit(metadata: Record<string, unknown> | undefined) {
+        assert(!!metadata, "Resubmitted ops must have metadata");
+        const { blobId, localId }: { blobId?: string; localId?: string; } = metadata;
+        if (!blobId) {
+            assert(!!localId, "Submitted BlobAttach ops must have a blobId or localId");
+            // We submitted this op while offline. The blob should have been uploaded by now.
+            const pendingEntry = this.pendingBlobs.get(localId);
+            assert(pendingEntry?.status === PendingBlobStatus.OfflinePendingOp &&
+                !!pendingEntry?.storageId, "blob must be uploaded before resubmitting BlobAttach op");
+            return this.sendBlobAttachOp(pendingEntry.storageId, localId);
+        }
+        return this.sendBlobAttachOp(blobId, localId);
     }
 
-    public processBlobAttachOp(blobId: string, local: boolean) {
+    public processBlobAttachOp(message: ISequencedDocumentMessage, local: boolean) {
+        assert(message?.metadata?.blobId, 0x12a /* "Missing blob id on metadata" */);
+        if (message.metadata.localId !== undefined) {
+            this.redirectTable.set(message.metadata.localId, message.metadata.blobId);
+        }
+        // set id. entry
+        this.redirectTable.set(message.metadata.blobId, message.metadata.blobId);
+
         if (local) {
-            const pendingBlobP = this.pendingBlobIds.get(blobId);
-            assert(pendingBlobP !== undefined, 0x1f8 /* "local BlobAttach op with no pending blob" */);
-            pendingBlobP.resolve();
-            this.pendingBlobIds.delete(blobId);
+            if (message.metadata.localId === undefined) {
+                for (const [id, entry] of this.pendingBlobs) {
```

Comment:
Yes, since there is no local ID it was sent while connected, but if we disconnect while waiting for it to round-trip we will submit another with the local ID, which we need to wait for instead

Reply:
I expanded the comment to capture my understanding.  Please have a look and fix as needed.

Reply:
Hmm, this comment is not accurate. I think you are mixing up localId and blobId/storageId. Since there is no local ID on this op, that means it was submitted while online/connected. I do submit ops with no blobId/storagId while offline, but since the client is disconnected, those won't reach process() until after they are passed to reSubmit(), which adds the storage ID. I will try to add some more comments to better explain this.

Thinking about it though, I realize we don't need to handle the case I was worried about here, since even if we transitioned the online blob to offline and submitted another op with the local ID, this op should guarantee that we won't need to reupload the blob.

I did avoid the linear search in a previous iteration of this PR by tracking pending online ops, but I figured it was not worth the added complexity since I don't expect to have many pending blobs at a time. I also do a linear search to find unuploaded offline blobs in onConnected() and online pending op blobs in onDisconnected(). pendingBlobs is not drained at this point because we only remove entries once the BlobAttach round-trips, and we only delay connection until they have been uploaded.